src/Microsoft.ComponentDetection.Detectors/pnpm/Pnpm9Detector.cs

```diff
@@ -0,0 +1,102 @@
+namespace Microsoft.ComponentDetection.Detectors.Pnpm;
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.ComponentDetection.Contracts;
+
+/// <summary>
+/// There is still no official docs for the new v9 lock if format, so these parsing contracts are empirically based.
+/// Issue tracking v9 specs: https://github.com/pnpm/spec/issues/6
+/// Format should eventually get updated here: https://github.com/pnpm/spec/blob/master/lockfile/6.0.md.
+/// </summary>
+public class Pnpm9Detector : IPnpmDetector
+{
+    public const string MajorVersion = "9";
+    private readonly PnpmParsingUtilitiesBase<PnpmYamlV9> pnpmParsingUtilities = PnpmParsingUtilitiesFactory.Create<PnpmYamlV9>();
+
+    public void RecordDependencyGraphFromFile(string yamlFileContent, ISingleFileComponentRecorder singleFileComponentRecorder)
+    {
+        var yaml = this.pnpmParsingUtilities.DeserializePnpmYamlFile(yamlFileContent);
+
+        // There may be multiple instance of the same package (even at the same version) in pnpm differentiated by other aspects of the pnpm dependency path.
+        // Therefor all DetectedComponents are tracked by the same full string pnpm uses, the pnpm dependency path, which is used as the key in this dictionary.
+        // Some documentation about pnpm dependency paths can be found at https://github.com/pnpm/spec/blob/master/dependency-path.md.
+        var components = new Dictionary<string, (DetectedComponent, Package)>();
+
+        // Create a component for every package referenced in the lock file.
+        // This includes all directly and transitively referenced dependencies.
+        foreach (var (pnpmDependencyPath, package) in yaml.Snapshots ?? Enumerable.Empty<KeyValuePair<string, Package>>())
+        {
+            // Ignore "file:" as these are local packages.
+            // Such local packages should only be referenced at the top level (via ProcessDependencyList) which also skips them or from other local packages (which this skips).
+            // There should be no cases where a non-local package references a local package, so skipping them here should not result in failed lookups below when adding all the graph references.
+            if (pnpmDependencyPath.StartsWith("file:"))
```

Comment:
nit: we can create a constant for this pnpm Dependency Path and share across all the versions that need it